### PR TITLE
Search vm snapshots using tags

### PR DESCRIPTION
**Pull request by Pearl1594**
## Description
Search VM snapshots using tags
Currently, search of VM snapshots doesn't comply with tags passed as input

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
 tested via CLI (cmk) and UI

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java [hunk 1]
```diff
@@ -229,11 +236,32 @@ public boolean stop() {
...
+            }
+            tagSearch.and("resourceType", tagSearch.entity().getResourceType(), SearchCriteria.Op.EQ);
+            sb.groupBy(sb.entity().getId());
+            sb.join("tagSearch", tagSearch, sb.entity().getId(), tagSearch.entity().getResourceId(), JoinBuilder.JoinType.INNER);
+        }
 
         SearchCriteria<VMSnapshotVO> sc = sb.create();
         _accountMgr.buildACLSearchCriteria(sc, domainId, isRecursive, permittedAccounts, listProjectResourcesCriteria);
 
+        if (tags != null && !tags.isEmpty()) {
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java:255]: `MapUtils.isNotEmpty` would fit nicely here

File: server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java [hunk 2]
```diff
@@ -229,11 +236,32 @@ public boolean stop() {
         sb.and("idIN", sb.entity().getId(), SearchCriteria.Op.IN);
         sb.and("display_name", sb.entity().getDisplayName(), SearchCriteria.Op.EQ);
         sb.and("account_id", sb.entity().getAccountId(), SearchCriteria.Op.EQ);
-        sb.done();
+
+        if (tags != null && !tags.isEmpty()) {
+            SearchBuilder<ResourceTagVO> tagSearch = _resourceTagDao.createSearchBuilder();
+            for (int count = 0; count < tags.size(); count++) {
+                tagSearch.or().op("key" + String.valueOf(count), tagSearch.entity().getKey(), SearchCriteria.Op.EQ);
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java:243]: A few values can be extracted to constants, especially considering that they are being used more than once, for instance `"key"`.

**Comment by rhtyd** [comment]
@Pearl1594 can you address review comments and kick packaging/tests - assuming this is for primate, if so - can you pl change the base branch to 4.14?

**Comment by Pearl1594** [comment]
@GabrielBrascher Thanks for reviewing, I've addressed your comments.
@rhtyd I've rebased it to 4.14
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@Pearl1594 pl address code review comments, send another PR for refactoring as necessary

**Pull request merged**